### PR TITLE
Fix ASAN / potential segfault in quantized Tensor memory allocations.

### PR DESCRIPTION
**Pull request by AshkanAliabadi**

### CODE REVIEW

**Comment by AshkanAliabadi** [comment]
@supriyar @jerryzh168 We must make sure all input tensors to QNNPACK are allocated with this allocator.

File: aten/src/ATen/quantized/Quantizer.cpp [hunk 1]
```diff
@@ -479,6 +480,43 @@ QTensorImpl* get_qtensorimpl(const Tensor& self) {
...
+// falls on an umapped page or one with no read permission.  Here we define a
+// custom allocator that allocates the extra storage required to keep this
+// behavior safe.
+struct QNNPACKAllocator final : at::Allocator {
+public:
+  virtual ~QNNPACKAllocator() override = default;
+
+  virtual at::DataPtr allocate(size_t nbytes) const override {
+    Cast memory{c10::alloc_cpu(kGuard + nbytes)};
+    memory.as_byte_ptr += kGuard;
```
Comment (supriyar) [inline aten/src/ATen/quantized/Quantizer.cpp]: Why do we need to do this here and below (in the deleter)? `memory.as_byte_ptr` doesn't seem to be used outside this struct.

Comment (AshkanAliabadi) [reply aten/src/ATen/quantized/Quantizer.cpp]: This is part of a union.  The variables memory.as_byte_ptr and memory.as_void_ptr occupy the same location in memory, hence modifying one affects the other.  Pointer arithmetic is not allowed on void pointers hence the need for this cast.

**Comment by AshkanAliabadi** [comment]
Actually coming to think about this even though QNNPACK only access beyond the bounds of input, the output tensor must be allocated with this allocator as well if it ends up being the input for a subsequent QNNPACK operation in a following layer, unless you guys are doing an output to input copy as part of required requantization step of some sort.

I am currently in the middle of some FP32 work and am gradually moving all mobile code to one location, this allocator included.  Afterwards we should unify our treatment of all these tensors with regards to memory allocation.

**Comment by supriyar** [comment]
> Actually coming to think about this even though QNNPACK only access beyond the bounds of input, the output tensor must be allocated with this allocator as well if it ends up being the input for a subsequent QNNPACK operation in a following layer, unless you guys are doing an output to input copy as part of required requantization step of some sort.
> 
> I am currently in the middle of some FP32 work and am gradually moving all mobile code to one location, this allocator included. Afterwards we should unify our treatment of all these tensors with regards to memory allocation.

Output tensors are allocated using `at::_empty_affine_quantized` inside each op, eg. https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qconv.cpp#L574. I would expect them to take the same path and call `new_qtensor_cpu`. @jerryzh168 is this accurate?

File: aten/src/ATen/quantized/Quantizer.cpp [hunk 2]
```diff
@@ -479,6 +480,43 @@ QTensorImpl* get_qtensorimpl(const Tensor& self) {
   return static_cast<QTensorImpl*>(self.unsafeGetTensorImpl());
 }
 
+// QNNPACK can access up to 8 bytes beyond the beginning of the tensor's storage
+// boundary which does trigger ASAN, and can result in a segfault if the memory
+// falls on an umapped page or one with no read permission.  Here we define a
+// custom allocator that allocates the extra storage required to keep this
+// behavior safe.
+struct QNNPACKAllocator final : at::Allocator {
```
Comment (supriyar) [inline aten/src/ATen/quantized/Quantizer.cpp]: Are we making this the default allocator for all qtensors? Should we use this only when the backend engine is set to qnnpack and use the default allocator for fbgemm?

File: aten/src/ATen/quantized/Quantizer.cpp [hunk 3]
```diff
@@ -487,16 +525,16 @@ inline Tensor new_qtensor_cpu(
   AT_ASSERT(options.device().is_cpu());
 
   native::check_size_nonnegative(sizes);
-  auto* allocator = at::getCPUAllocator();
+  static QNNPACKAllocator allocator;
```
Comment (jerryzh168) [inline aten/src/ATen/quantized/Quantizer.cpp]: what about non-qnnpack environment?

File: aten/src/ATen/quantized/Quantizer.cpp (hunk 2 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/quantized/Quantizer.cpp]: This should be the default allocator for any input tensor that ends up in QNNPACK.

**Comment by jerryzh168** [comment]
> > Actually coming to think about this even though QNNPACK only access beyond the bounds of input, the output tensor must be allocated with this allocator as well if it ends up being the input for a subsequent QNNPACK operation in a following layer, unless you guys are doing an output to input copy as part of required requantization step of some sort.
> > I am currently in the middle of some FP32 work and am gradually moving all mobile code to one location, this allocator included. Afterwards we should unify our treatment of all these tensors with regards to memory allocation.
> 
> Output tensors are allocated using `at::_empty_affine_quantized` inside each op, eg. https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qconv.cpp#L574. I would expect them to take the same path and call `new_qtensor_cpu`. @jerryzh168 is this accurate?

yeah `_empty_affine_quantized` is calling `new_qtensor_cpu` under the hood

File: aten/src/ATen/quantized/Quantizer.cpp (hunk 3 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/quantized/Quantizer.cpp]: You mean FBGEMM uses this code as well?  In that case we shouldn't be using this allocator.  That is decided by a compile-time flag, right?

File: aten/src/ATen/quantized/Quantizer.cpp (hunk 2 above)
Comment (supriyar) [reply aten/src/ATen/quantized/Quantizer.cpp]: All qtensors will take this path, so we should only use this when the backend is set to QNNPACK.

File: aten/src/ATen/quantized/Quantizer.cpp (hunk 3 above)
Comment (jerryzh168) [reply aten/src/ATen/quantized/Quantizer.cpp]: this should probably be guarded by qengine and compiler flag as well if this is only compilable on qnnpack I think.

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 3252a8a7:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=7078f4b27ddfc5752c7e1601488feac89deec601&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 7078f4b2 since Jan 02
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  Since your merge base is older than `viable/strict`, run these commands:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=3252a8a74f0471a529386797e3d33e60254a3730).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_macos_10_13_py3_test](https://dr.pytorch.org/build-details.html?build_id=79677164) from Jan 02 until Jan 02



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: aten/src/ATen/quantized/Quantizer.cpp (hunk 2 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/quantized/Quantizer.cpp]: Thanks for pointing that out Jerry.  This should be OK for FBGEMM as well.

File: aten/src/ATen/quantized/Quantizer.cpp [hunk 4]
```diff
@@ -487,16 +525,16 @@ inline Tensor new_qtensor_cpu(
   AT_ASSERT(options.device().is_cpu());
 
   native::check_size_nonnegative(sizes);
-  auto* allocator = at::getCPUAllocator();
+  static QAllocator allocator;
```
Comment (jerryzh168) [inline aten/src/ATen/quantized/Quantizer.cpp]: is this allocator only used for qnnpack? I think you'll need to make sure the fbgemm will use the original CPU allocator?

Comment (AshkanAliabadi) [reply aten/src/ATen/quantized/Quantizer.cpp]: It should work for FBGEMM as well at the cost of wasting 8 bytes per allocation.  Do you prefer keeping FBGEMM's code path intact?

Comment (jerryzh168) [reply aten/src/ATen/quantized/Quantizer.cpp]: yeah, I think it makes more sense to keep the original code path, you can use qengine to distinguish between these cases

Comment (jerryzh168) [reply aten/src/ATen/quantized/Quantizer.cpp]:
```
if (QNNPACK) {
// use qnnpack allocator
} else {
// use the original allocator
}
```

Comment (AshkanAliabadi) [reply aten/src/ATen/quantized/Quantizer.cpp]: Thanks Jerry.  Will upload a patch with these changes shortly.

Comment (supriyar) [reply aten/src/ATen/quantized/Quantizer.cpp]: I agree that we should keep FBGEMM separate. Please use the complier flag as well as `qengine` to use the new allocator for QNNPACK.

File: aten/src/ATen/quantized/Quantizer.cpp [hunk 5]
```diff
@@ -479,6 +480,43 @@ QTensorImpl* get_qtensorimpl(const Tensor& self) {
   return static_cast<QTensorImpl*>(self.unsafeGetTensorImpl());
 }
 
+// QNNPACK can access up to 8 bytes beyond the beginning of the tensor's storage
+// boundary which does trigger ASAN, and can result in a segfault if the memory falls
+// on a different page out of the process's address space.
+// Here we define a custom allocator that allocates the extra storage required to keep
+// this behavior safe.  This same allocator can be used for FBGEMM as well.
+struct QAllocator final : at::Allocator {
```
Comment (raghuramank100) [inline aten/src/ATen/quantized/Quantizer.cpp]: Can we have a test for this?

**Review by jerryzh168 — APPROVED** [review]
LGTM

File: aten/src/ATen/quantized/Quantizer.cpp [hunk 6]
```diff
@@ -1,6 +1,7 @@
 #include <ATen/ATen.h>
 #include <ATen/quantized/Quantizer.h>
 #include <c10/core/Allocator.h>
+#include <c10/core/CPUAllocator.h>
```
Comment (jerryzh168) [inline aten/src/ATen/quantized/Quantizer.cpp]: do we need this import? looks like `at::getCPUAllocator()` works before

Comment (AshkanAliabadi) [reply aten/src/ATen/quantized/Quantizer.cpp]: Thanks, removed.

Comment (AshkanAliabadi) [reply aten/src/ATen/quantized/Quantizer.cpp]: Apparently it's needed.

Comment (jerryzh168) [reply aten/src/ATen/quantized/Quantizer.cpp]: Oh, I see, it's for `c10::alloc_cpu` I think.

**Pull request closed without merging**